packages/zoltan2/core/src/algorithms/color/Zoltan2_AlgHybridD2.hpp

```diff
@@ -0,0 +1,416 @@
+
+
+namespace Zoltan2{
+
+template <typename Adapter>
+class AlgDistance2 : public AlgTwoGhostLayer<Adapter> {
+
+  public:
+    
+    using lno_t = typename Adapter::lno_t;
+    using gno_t = typename Adapter::gno_t;
+    using offset_t = typename Adapter::offset_t;
+    using scalar_t = typename Adapter::scalar_t;
+    using base_adapter_t = typename Adapter::base_adapter_t;
+    using map_t = Tpetra::Map<lno_t,gno_t>;
+    using femv_scalar_t = int;
+    using femv_t = Tpetra::FEMultiVector<femv_scalar_t, lno_t, gno_t>; 
+    using device_type = Tpetra::Map<>::device_type;//Kokkos::Device<Kokkos::Cuda, Kokkos::Cuda::memory_space>;
+    using execution_space = Tpetra::Map<>::execution_space;//Kokkos::Cuda;
+    using memory_space = Tpetra::Map<>::memory_space;//Kokkos::Cuda::memory_space;
+  private:
+    
+    virtual void colorInterior(const size_t nVtx,
+                       Kokkos::View<lno_t*,device_type> adjs_view,
+                       Kokkos::View<offset_t*, device_type> offset_view,
+                       Teuchos::RCP<femv_t> femv,
+		       Kokkos::View<lno_t*, device_type> vertex_list,
+		       size_t vertex_list_size = 0,
+                       bool recolor=false){
+      using KernelHandle = KokkosKernels::Experimental::KokkosKernelsHandle
+          <offset_t, lno_t, lno_t, execution_space, memory_space, memory_space>;
+  
+      KernelHandle kh;
+      kh.set_verbose(this->verbose); 
+      if(recolor)kh.create_distance2_graph_coloring_handle(KokkosGraph::COLORING_D2_VB_BIT);
+      else kh.create_distance2_graph_coloring_handle(KokkosGraph::COLORING_D2_NB_BIT);
+      if(vertex_list_size != 0){
+        kh.get_distance2_graph_coloring_handle()->set_vertex_list(vertex_list, vertex_list_size);
+      }
+      Kokkos::View<int**, Kokkos::LayoutLeft> femvColors = femv->template getLocalView<memory_space>();
+      Kokkos::View<int*, Kokkos::Device<execution_space,memory_space>> sv = subview(femvColors, Kokkos::ALL, 0);
+      kh.get_distance2_graph_coloring_handle()->set_vertex_colors(sv);
+
+      KokkosGraph::Experimental::graph_color_distance2(&kh, nVtx, offset_view, adjs_view);
+      this->numColors = kh.get_distance2_graph_coloring_handle()->get_num_colors();
+      if(this->verbose) std::cout<<"\nKokkosKernels Coloring: "<<kh.get_distance2_graph_coloring_handle()->get_overall_coloring_time()<<"\n";
+    
+    }
+    
+    virtual void colorInterior_serial(const size_t nVtx,
+		       typename Kokkos::View<lno_t*, device_type>::HostMirror adjs_view,
+		       typename Kokkos::View<offset_t*, device_type>::HostMirror offset_view,
+		       Teuchos::RCP<femv_t> femv,
+		       typename Kokkos::View<lno_t*, device_type>::HostMirror vertex_list,
+		       size_t vertex_list_size = 0,
+		       bool recolor=false){
+      using KernelHandle = KokkosKernels::Experimental::KokkosKernelsHandle
+          <offset_t, lno_t, lno_t, Kokkos::DefaultHostExecutionSpace, memory_space, memory_space>;
```

Comment:
memory_space will be device memory.  You want host memory here, right?  This code will likely fail without UVM.